### PR TITLE
Refactored QueryRunner::collectData

### PR DESCRIPTION
**Pull request by rmclaren**
## Description

This pull request refactors the data collection process in QueryRunner::collectData. The user benefits from a small performance improvement in the data collection process, but the main benefit is much easier to maintain code.

The C interface on NCEPLIB-bufr was extended to expose a data structure that makes it possible to determine the number of repeats in a fixed repeated sequence. This means we don't need to walk the data structure any more to discover these values. Instead we can get the counts directly.

A new class called NodeLookupTable was introduced. So for each node we care about, given the targets we are looking for (these include the target nodes (ex: TMBR) plus the intermediate nodes (like BRIT) for those targets. Its initialized with a data provider and the targets, and it collects the relevant data from the currently open message subset. 

** NOTE: This pull request requires a new NCEPLIB-bufr tag to work **bufr_v12.0.0**.

### Issue(s) addressed

Is a part of #1181 

## Acceptance Criteria (Definition of Done)

Unit tests run successfully.

## Dependencies

Has the following updated external dependency.
NCEPLIB-bufr  [bufr_v12.0.0]. https://github.com/NOAA-EMC/NCEPLIBS-bufr/projects/4#card-87633084

### CODE REVIEW

**Comment by rmclaren** [comment]
No idea why the commit list is so long... Doesn't make sense.

**Comment by BenjaminRuston** [comment]
@climbfuji adding you (or someone from INFRA) as well as INFRA label as there is a dependency on the NCEP BUFRlib which would need to be in the stack

**Comment by climbfuji** [comment]
> @climbfuji adding you (or someone from INFRA) as well as INFRA label as there is a dependency on the NCEP BUFRlib which would need to be in the stack

Thanks for the heads up @BenjaminRuston ! Once bufrlib 12.0.0 is available, please have someone create an issue in spack-stack to update the version and we'll take it from there.

**Comment by PatNichols** [comment]
@rmclaren @BenjaminRuston @climbfuji  After checking the recent releases for NCEP-bufrlib I found that 11.7.x was the latest release. Is 12.0 a development branch? I am somewhat confused on how we are to get this version.

**Comment by rmclaren** [comment]
No idea why the tests are failing as I don't have access to the logs...

**Comment by srherbener** [comment]
@rmclaren please try clicking on the CDASH links above. These are the lines in the check report above that have green check marks and titles like `AWS CodeBuild gnu - link to the corresponding CDASH page for viewing...`.

The CDASH links should take you to a web page (of which you should have access) that summarizes the test failures. For the GNU case, I see this message:
```
/jcsda/ioda-bundle/iodaconv/src/bufr/BufrParser/Query/DataProvider/NcepDataProvider.cpp: In member function 'virtual void Ingester::bufr::NcepDataProvider::updateTableData(const string&)':
/jcsda/ioda-bundle/iodaconv/src/bufr/BufrParser/Query/DataProvider/NcepDataProvider.cpp:69:13: error: 'get_irf_f' was not declared in this scope; did you mean 'get_inv_f'?
   69 |             get_irf_f(&intPtr, &size);
      |             ^~~~~~~~~
      |             get_inv_f
```
If you can't access the CDASH web page, please let us know and we will get that resolved.

**Comment by rmclaren** [comment]
That makes sense actually, its because they are using the older version of NCEPLIB-bufr (not v12 yet). 

I time out trying to access the CDASH pages unfortunately...

**Comment by srherbener** [comment]
@rmclaren, thanks for the explanation.

I'll ask about CDASH. It should be public. Do you think it might be an issue with a firewall on your side?

**Comment by rmclaren** [comment]
CDASH seems to be working now. Pretty sure its not a firewall on my side...

**Comment by rmclaren** [comment]
@climbfuji This PR already moves to to the **bufr_4** version of the NCEPLIB-bufr 12 library. I don't know if there is a need for an issue for this...

**Review by emilyhcliu — APPROVED** [review]
Tested and approved!

**Comment by PatNichols** [comment]
@rmclaren looks like we finally are able to test this!

File: src/ncar-bufr2nc-fortran/CMakeLists.txt [hunk 1]
```diff
@@ -21,4 +21,4 @@ ecbuild_add_executable( TARGET  bufr2nc_fortran.x
                         SOURCES ${bufr2nc_source}
                         LIBS    bufr::bufr_4
                                 NetCDF::NetCDF_Fortran
-				iodaconv_utils)
+				                        iodaconv_utils)
```
Comment (PatNichols) [inline src/ncar-bufr2nc-fortran/CMakeLists.txt:24]: @rmclaren could you delete the space here.

Comment (rmclaren) [reply src/ncar-bufr2nc-fortran/CMakeLists.txt]: @PatNichols Done

**Review by srherbener — APPROVED** [review]
Looks good! I built and tested on my Mac M1 (arm64) and I'm seeing  tests fail, but they match test failures that I see using the develop branch. I'll submit an issue for the develop test failures on the Mac M1 (these test failures seem to occur only on the Mac M1).

If it's possible, please break up PRs into smaller sets of changes. It can be difficult and time consuming to review a large set of changes. I realize that this isn't possible sometimes, so no worries now and please keep in mind for future PRs. Thanks!

**Comment by rmclaren** [comment]
@srherbener Thanks for looking at this.

**Pull request merged**